# LAB 02 - Làm quen với PyTorch

**Họ và tên:** NgoNguyenThanhNhan  
**Mã sinh viên:** 3123411205

Notebook đã chạy với PyTorch và có đầu ra kèm theo cho từng bài.


## Kết quả thực hành


### Bài 1 Tạo Tensor cơ bản


In [1]:
import torch
torch.manual_seed(3123411205)

a = torch.tensor([[1, 2, 3], [4, 5, 6]], dtype=torch.int32)
print("Tensor a:\n", a)
print("Kích thước:", a.shape)
print("Số chiều:", a.ndim)
print("Kiểu dữ liệu:", a.dtype)


Tensor a:
 tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)
Kích thước: torch.Size([2, 3])
Số chiều: 2
Kiểu dữ liệu: torch.int32


**Nhận xét / trả lời:** Tensor a có kích thước 2 x 3, gồm 2 chiều và dùng kiểu số nguyên torch.int32.


### Bài 2 Tạo các Tensor thường dùng


In [2]:
a = torch.rand(3, 4)
b = torch.randn(3, 4)
c = torch.zeros(2, 3)
d = torch.ones(2, 3)
e = torch.eye(4)
f = torch.full((5, 5), 10)
print("rand:\n", a)
print("randn:\n", b)
print("zeros:\n", c)
print("ones:\n", d)
print("eye:\n", e)
print("Tensor 5 x 5 toàn 10:\n", f)


rand:
 tensor([[0.7134, 0.3682, 0.7664, 0.4444],
        [0.3401, 0.5112, 0.9183, 0.7149],
        [0.2726, 0.5733, 0.3032, 0.3273]])
randn:
 tensor([[-1.6963,  0.4559,  0.4894,  0.1482],
        [ 0.2816,  1.1282, -0.4856,  1.4614],
        [ 0.4290,  0.4906, -0.5857,  0.0110]])
zeros:
 tensor([[0., 0., 0.],
        [0., 0., 0.]])
ones:
 tensor([[1., 1., 1.],
        [1., 1., 1.]])
eye:
 tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])
Tensor 5 x 5 toàn 10:
 tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


**Nhận xét / trả lời:** torch.rand() sinh các giá trị phân bố đều trong khoảng [0, 1). torch.randn() sinh các giá trị theo phân bố chuẩn chuẩn hóa, nên có thể âm hoặc lớn hơn 1. Tensor f được tạo bằng torch.full((5, 5), 10).


### Bài 3 Kiểm tra thông tin Tensor


In [3]:
a = torch.zeros(2, 3, 4)
print("Tensor:\n", a)
print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)


Tensor:
 tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions: 3
Data type: torch.float32


**Nhận xét / trả lời:** torch.Size([2, 3, 4]) cho biết Tensor có 3 chiều: chiều đầu có 2 phần tử, chiều thứ hai có 3 phần tử và chiều cuối có 4 phần tử. Tổng cộng Tensor có 2 x 3 x 4 = 24 phần tử.


### Bài 4 Slicing và Indexing


In [4]:
a = torch.randn(3, 4, 5)
print("Original tensor:\n", a)
print("\na[1]:\n", a[1])
print("\na[1:, 2:4]:\n", a[1:, 2:4])
print("\nHai phần tử cuối ở chiều thứ ba a[:, :, -2:]:\n", a[:, :, -2:])
print("Shape a[1]:", a[1].shape)
print("Shape a[1:, 2:4]:", a[1:, 2:4].shape)
print("Shape a[:, :, -2:]:", a[:, :, -2:].shape)


Original tensor:
 tensor([[[-0.7104,  0.2812,  0.9728,  1.0132, -0.9108],
         [-1.1050, -1.4213,  0.9957, -0.4083,  1.0303],
         [-1.0517,  0.9607,  1.3560, -1.1295, -0.5569],
         [-0.3061,  0.1009, -0.2236,  0.9509,  0.7644]],

        [[ 1.2238,  2.0194, -1.6934,  2.6144, -0.1781],
         [ 0.5437, -0.4105,  1.1784, -1.2401, -0.2386],
         [ 1.0049, -0.9259, -0.7361, -0.9919, -1.0333],
         [-0.5816,  0.1065,  0.1962,  0.4147,  3.3501]],

        [[ 0.9469, -1.4126,  1.4123, -0.5493, -1.7476],
         [ 0.9598, -0.3961,  0.0943, -2.2999,  0.7534],
         [ 0.2388, -0.9953,  0.8920, -0.6846,  1.8046],
         [ 1.6823,  1.3632, -1.1528,  1.3316,  0.0491]]])

a[1]:
 tensor([[ 1.2238,  2.0194, -1.6934,  2.6144, -0.1781],
        [ 0.5437, -0.4105,  1.1784, -1.2401, -0.2386],
        [ 1.0049, -0.9259, -0.7361, -0.9919, -1.0333],
        [-0.5816,  0.1065,  0.1962,  0.4147,  3.3501]])

a[1:, 2:4]:
 tensor([[[ 1.0049, -0.9259, -0.7361, -0.9919, -1.0333],
     

**Nhận xét / trả lời:** a[1] lấy lát cắt tại chỉ số đầu tiên 1 nên còn shape (4, 5). a[1:, 2:4] lấy 2 phần tử ở chiều 1 và 2 phần tử ở chiều 2 nên có shape (2, 2, 5). Biểu thức a[:, :, -2:] lấy hai phần tử cuối của chiều thứ ba, có shape (3, 4, 2).


### Bài 5 Reshape và Flatten


In [5]:
a = torch.randn(3, 4)
print("Original:\n", a)
print("\nFlatten:\n", a.ravel())
print("\nReshape 3 x 2 x 2:\n", a.reshape(3, 2, 2))
print("\nReshape 2 x 6:\n", a.reshape(2, 6))
print("Số phần tử trước:", a.numel())
print("Số phần tử sau:", a.reshape(2, 6).numel())


Original:
 tensor([[-0.1563,  0.0414, -1.4019, -1.7382],
        [-1.0688, -0.7074, -1.0383, -0.4099],
        [ 0.8185,  0.0774,  0.1145, -0.5713]])

Flatten:
 tensor([-0.1563,  0.0414, -1.4019, -1.7382, -1.0688, -0.7074, -1.0383, -0.4099,
         0.8185,  0.0774,  0.1145, -0.5713])

Reshape 3 x 2 x 2:
 tensor([[[-0.1563,  0.0414],
         [-1.4019, -1.7382]],

        [[-1.0688, -0.7074],
         [-1.0383, -0.4099]],

        [[ 0.8185,  0.0774],
         [ 0.1145, -0.5713]]])

Reshape 2 x 6:
 tensor([[-0.1563,  0.0414, -1.4019, -1.7382, -1.0688, -0.7074],
        [-1.0383, -0.4099,  0.8185,  0.0774,  0.1145, -0.5713]])
Số phần tử trước: 12
Số phần tử sau: 12


**Nhận xét / trả lời:** reshape và ravel chỉ thay đổi cách nhìn dữ liệu, không thay đổi số phần tử. Tensor ban đầu có 3 x 4 = 12 phần tử; cả dạng 3 x 2 x 2 và 2 x 6 đều có 12 phần tử.


### Bài 6 Phép toán và nhân ma trận


In [6]:
a = torch.randn(2, 3)
b = torch.randn(2, 3)
print("a:\n", a)
print("b:\n", b)
print("\na + b:\n", a + b)
print("\na * b:\n", a * b)
print("\na / b:\n", a / b)
print("\na^2:\n", a ** 2)
print("\nMatrix multiplication a @ b.T:\n", a @ b.T)
print("Shape a @ b.T:", (a @ b.T).shape)


a:
 tensor([[-1.2167, -0.2092, -0.3123],
        [ 1.3746,  0.8801, -2.2775]])
b:
 tensor([[ 0.7240,  0.4720,  1.4067],
        [ 0.4920, -0.1284, -0.3133]])

a + b:
 tensor([[-0.4928,  0.2628,  1.0944],
        [ 1.8666,  0.7516, -2.5908]])

a * b:
 tensor([[-0.8809, -0.0987, -0.4393],
        [ 0.6763, -0.1130,  0.7136]])

a / b:
 tensor([[-1.6806, -0.4431, -0.2220],
        [ 2.7939, -6.8534,  7.2682]])

a^2:
 tensor([[1.4805, 0.0437, 0.0975],
        [1.8896, 0.7745, 5.1868]])

Matrix multiplication a @ b.T:
 tensor([[-1.4189, -0.4739],
        [-1.7932,  1.2770]])
Shape a @ b.T: torch.Size([2, 2])


**Nhận xét / trả lời:** a * b là phép nhân từng phần tử nên kết quả vẫn có shape (2, 3). a @ b.T là nhân ma trận: (2, 3) nhân (3, 2), do đó kết quả có shape (2, 2).


### Bài 7 Các hàm thống kê trên Tensor


In [7]:
a = torch.randn(3, 4)
print("Tensor a:\n", a)
print("\nMean theo cột dim=0:\n", torch.mean(a, dim=0))
print("\nStandard deviation theo cột dim=0:\n", torch.std(a, dim=0))
print("\nCumulative sum theo cột dim=0:\n", torch.cumsum(a, dim=0))
print("\nMean theo hàng dim=1:\n", torch.mean(a, dim=1))
print("\nStandard deviation theo hàng dim=1:\n", torch.std(a, dim=1))
print("\nCumulative sum theo hàng dim=1:\n", torch.cumsum(a, dim=1))


Tensor a:
 tensor([[ 1.0814, -0.1541, -2.1701,  0.5156],
        [ 0.9388,  1.7984, -0.6956, -1.4383],
        [-1.0882,  1.3724, -1.3355,  0.2679]])

Mean theo cột dim=0:
 tensor([ 0.3107,  1.0056, -1.4004, -0.2183])

Standard deviation theo cột dim=0:
 tensor([1.2135, 1.0266, 0.7394, 1.0638])

Cumulative sum theo cột dim=0:
 tensor([[ 1.0814, -0.1541, -2.1701,  0.5156],
        [ 2.0202,  1.6443, -2.8656, -0.9227],
        [ 0.9320,  3.0167, -4.2012, -0.6548]])

Mean theo hàng dim=1:
 tensor([-0.1818,  0.1508, -0.1958])

Standard deviation theo hàng dim=1:
 tensor([1.4184, 1.4806, 1.2609])

Cumulative sum theo hàng dim=1:
 tensor([[ 1.0814,  0.9273, -1.2427, -0.7272],
        [ 0.9388,  2.7372,  2.0416,  0.6033],
        [-1.0882,  0.2842, -1.0513, -0.7834]])


**Nhận xét / trả lời:** dim=0 gom các phần tử cùng cột, vì vậy mean và std trả về 4 giá trị; cumsum cộng dồn từ trên xuống. dim=1 gom các phần tử cùng hàng, nên trả về 3 giá trị; cumsum cộng dồn từ trái sang phải.


### Bài 8 Autograd cơ bản


In [8]:
x = torch.tensor(3.6, requires_grad=True)
y = x * x
y.backward()
print("Với y = x^2")
print("x =", x.item())
print("y =", y.item())
print("x.grad =", x.grad.item())
print("Đạo hàm thủ công 2x =", 2 * x.item())

x2 = torch.tensor(3.6, requires_grad=True)
y2 = 3 * x2**2 + 2 * x2 + 1
y2.backward()
print("\nVới y = 3x^2 + 2x + 1")
print("y =", y2.item())
print("x.grad =", x2.grad.item())
print("Đạo hàm thủ công 6x + 2 =", 6 * x2.item() + 2)


Với y = x^2
x = 3.5999999046325684
y = 12.959999084472656
x.grad = 7.199999809265137
Đạo hàm thủ công 2x = 7.199999809265137

Với y = 3x^2 + 2x + 1
y = 47.07999801635742
x.grad = 23.599998474121094
Đạo hàm thủ công 6x + 2 = 23.59999942779541


**Nhận xét / trả lời:** Với y = x^2, dy/dx = 2x; tại x = 3.6 thì bằng 7.2, trùng với x.grad. Với y = 3x^2 + 2x + 1, dy/dx = 6x + 2; tại x = 3.6 bằng 23.6, trùng với gradient do autograd tính.


### Bài 9 Gradient Descent tìm hệ số đa thức


In [11]:
import numpy as np
import torch
np.random.seed(3123411205)
polynomial = np.poly1d([1, 2, 3])
n = 40
X = np.random.randn(n, 1) * 5
Y = polynomial(X)
XX = np.hstack([X * X, X, np.ones_like(X)])

w = torch.randn(3, 1, requires_grad=True)
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)
optimizer = torch.optim.NAdam([w], lr=0.03)
print("Initial coefficients:", w.detach().flatten().numpy())
for epoch in range(2500):
    optimizer.zero_grad()
    y_pred = x @ w
    mse = torch.mean((y - y_pred) ** 2)
    mse.backward()
    optimizer.step()
print("Learned coefficients:", w.detach().flatten().numpy())
print("Target coefficients:", np.array([1., 2., 3.]))
print("Final MSE:", mse.item())


Initial coefficients: [0.804909   0.10706162 1.5187664 ]
Learned coefficients: [1.        1.9999994 2.999999 ]
Target coefficients: [1. 2. 3.]
Final MSE: 1.018349814196906e-11


**Nhận xét / trả lời:** Các hệ số học được xấp xỉ [1, 2, 3], là hệ số đã dùng để sinh dữ liệu. optimizer.zero_grad() xóa gradient tích lũy từ vòng lặp trước; mse.backward() lan truyền ngược để tính gradient của MSE theo w; optimizer.step() cập nhật w theo thuật toán NAdam để giảm MSE.


### Bài 10 Giải hệ phương trình bằng Autograd


In [ ]:
random.seed(3123411205)
A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)
optimizer = torch.optim.NAdam([A, B, C, D], lr=0.03)

for _ in range(4000):
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2
    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4
    optimizer.zero_grad()
    sqerr.backward()
    optimizer.step()

print(f"A = {A.item():.6f}")
print(f"B = {B.item():.6f}")
print(f"C = {C.item():.6f}")
print(f"D = {D.item():.6f}")
print(f"A + B = {(A+B).item():.6f}")
print(f"C - D = {(C-D).item():.6f}")
print(f"A + C = {(A+C).item():.6f}")
print(f"B - D = {(B-D).item():.6f}")
print(f"Tổng sai số bình phương = {sqerr.item():.10f}")


**Nhận xét / trả lời:** Các giá trị thay vào bốn phương trình cho kết quả xấp xỉ 9, 1, 8 và 2. Hệ có vô số nghiệm vì một phương trình phụ thuộc các phương trình còn lại: từ A+B=9, A+C=8 và B-D=2 suy ra C-D=1. Có thể đặt A=t, khi đó B=9-t, C=8-t và D=7-t.


### Bài 11 Bài tập áp dụng


In [ ]:
X = torch.arange(24)
print("X reshape 4 x 6:\n", X.reshape(4, 6))
print("\nX reshape 2 x 12:\n", X.reshape(2, 12))
print("\nX reshape 2 x 3 x 4:\n", X.reshape(2, 3, 4))

A = torch.randn(3, 4)
B = torch.randn(3, 4)
print("\nA + B:\n", A + B)
print("\nA * B:\n", A * B)
print("\nA / B:\n", A / B)
print("\nTrung bình theo cột:\n", A.mean(dim=0))
print("\nĐộ lệch chuẩn theo cột:\n", A.std(dim=0))

x = torch.tensor(2.0, requires_grad=True)
y = 2*x**2 + 5*x + 3
y.backward()
print("\ny tại x=2:", y.item())
print("dy/dx từ autograd:", x.grad.item())
print("dy/dx thủ công 4x + 5:", 4*x.item() + 5)

np.random.seed(1205)
X_train = np.random.randn(50, 1) * 4
Y_train = 2*X_train**2 - 3*X_train + 5
design = np.hstack([X_train**2, X_train, np.ones_like(X_train)])
w = torch.randn(3, 1, requires_grad=True)
design_t = torch.tensor(design, dtype=torch.float32)
target_t = torch.tensor(Y_train, dtype=torch.float32)
optimizer = torch.optim.NAdam([w], lr=0.03)
for epoch in range(3000):
    optimizer.zero_grad()
    loss = torch.mean((design_t @ w - target_t) ** 2)
    loss.backward()
    optimizer.step()
print("\nHệ số học được cho y = 2x^2 - 3x + 5:", w.detach().flatten().numpy())
print("MSE cuối:", loss.item())


**Nhận xét / trả lời:** X gồm 24 phần tử nên có thể reshape hợp lệ thành 4 x 6, 2 x 12 và 2 x 3 x 4. Với y = 2x^2 + 5x + 3, dy/dx = 4x + 5; tại x=2 có giá trị 13, trùng với autograd. Sau huấn luyện, các hệ số của đa thức mới xấp xỉ [2, -3, 5].
